pymatgen/core/grain_boundary_generator.py

```diff
@@ -0,0 +1,1647 @@
+    def gb_from_matrices(self, top_grain_matrix, bottom_grain_matrix, expand_times, vacuum_thickness):
+
+        """
+        This function is used to generate GB structures from the given transformation matrices.
+        The idea here is to use a generic descriptor for a specific type of GB define by
+        spacegroup and the matrix only. This descriptor will then work for any material of
+        a particular spacegroup.
+        E.g. for sigma 5 (001) fcc GB, we know beforehand matrix to describe the two
+        grains of this GB, which can be used to generate GB structure for all kinds of systems
+        with fcc lattice. To be used for the high throughput effort.
+
+        Args:
+           top_grain_matrix (3 by 3 array): The transformation matrix used to operate on top grain.
+           bottom_grain_matrix (3 by 3 array): The transformation matrix used to operate on bottom grain.
+           expand_times (int): The multiple times used to expand one unit grain to larger grain.
+               This is used to tune the grain length of GB to warrant that the two GBs in one cell do
+               not interact with each other.
+           vacuum_thickness (float): The thickness of vacuum that you want to insert between
+                two grains of the GB.
+        Returns:
+           Grain boundary structure.
+        """
+
+        parent_structure = self.initial_structure.copy()
+        # top grain
+        top_grain = fix_pbc(parent_structure * top_grain_matrix)
+        top_grain_init = top_grain.lattice.matrix
+        top_grain.make_supercell([1, 1, expand_times])
+        top_grain = fix_pbc(top_grain)
+
+        # bottom grain, using top grain's lattice matrix
+        bottom_grain = fix_pbc(parent_structure * bottom_grain_matrix, top_grain_init)
+        bottom_grain.make_supercell([1, 1, expand_times])
+        bottom_grain = fix_pbc(bottom_grain, top_grain.lattice.matrix)
+
+        # construct all species
+        all_species = []
+        all_species.extend([site.specie for site in bottom_grain])
+        all_species.extend([site.specie for site in top_grain])
+
+        half_lattice = top_grain.lattice
+        # calculate translation vector, perpendicular to the plane
+        normal_v_plane = np.cross(half_lattice.matrix[0], half_lattice.matrix[1])
+        unit_normal_v = normal_v_plane / np.linalg.norm(normal_v_plane)
+        translation_v = unit_normal_v * vacuum_thickness
+
+        # construct the final lattice
+        whole_matrix_no_vac = half_lattice.matrix
+        whole_matrix_no_vac[2] = half_lattice.matrix[2] * 2
+        whole_matrix_with_vac = whole_matrix_no_vac.copy()
+        whole_matrix_with_vac[2] = whole_matrix_no_vac[2] + translation_v * 2
+        whole_lat = Lattice(whole_matrix_with_vac)
+
+        # construct the coords, move top grain with translation_v
+        all_coords = []
+        for site in bottom_grain:
+            all_coords.append(site.coords)
+        for site in top_grain:
```

Comment:
It is easy to label the grain sites( top or bottom, already implemented). 
For the co-incident sites, I do not have a good way to label them for now.
Welcome for any suggestions.

Reply:
Many thanks for adding grain labels.

For coincident sites, is it possible to create an additional structure where you extend both the top and bottom grains to fill the entire cell?

Reply:
You mean one structure with twice top grain and twice bottom grain. e.g. both top grain and bottom grain have N atoms. The grain boundary structure has 2N atoms. The new structure has 4N atoms with some atoms overlapping with each other. Is that correct?

Reply:
Matt, you approach should work, thank you very much! Xiang-Guo concerned about the speed after expanding the cell, but he thinks of a way that we could label the co-incident sites before expand to the larger GB and do the comparison, then expand the unit grain boundary to larger grain boundary and the properties of the labeled incident sites are also repeated. We are testing now and Xiang-Guo will implement it very soon. Thank you all, great minds!

Reply:
Hi Matt, I had a version that can label the co-incident sites. Please check if that what you want.

Reply:
Just tested it, looks like it works great! :)

Here's a quick prototype visualization:

<img width="842" alt="image" src="https://user-images.githubusercontent.com/2976580/42606628-9078d542-8533-11e8-8d96-372a2c6ad1e4.png">

Reply:
(Here's a better angle)

<img width="546" alt="image" src="https://user-images.githubusercontent.com/2976580/42608197-95810bc4-853b-11e8-9ef4-d2de8c5eff9d.png">

Reply:
Hi Matt, that's a pretty visualization~~ I guess the atoms with yellow and purple colors are the coincidence sites? Do you think using one color for all the coincidence sites would be easier for the users to locate these coincidence sites? For me, the first figure is actually more obvious to recognize these coincidence sites, since they line up across two grains. Thank you for your efficient work!

Reply:
Completely agreed, these were just the default colours don’t worry. We can change them to something more sensible!

Reply:
Oh yeah, the second graph is easier to recognize the two grains has a rotation angle. Also, I am wondering will it be difficult to add the icons which allow the users to rotate around specific directions (e.g.: x, y, and z-axis or rotation axis in a more general way) like the icons that are available in Vesta? Which might be convenient for users to look through the characteristics of GB structures by rotate it along the rotation axis.

Reply:
This is fully 3D so you can rotate it :) If you’re on the update call tomorrow, I’ll give a demonstration.

Reply:
Oh, it's Friday! Even better. Rotating about an axis should be fine, as long as I have the axis of rotation in the Cartesian space of the crystal. I'll take a look tomorrow and see if I can make it work.